In [1]:
%load_ext watermark
%watermark

2018-12-30T16:48:56-06:00

CPython 3.6.7
IPython 7.2.0

compiler   : GCC 8.2.0
system     : Linux
release    : 4.15.0-43-generic
machine    : x86_64
processor  : x86_64
CPU cores  : 4
interpreter: 64bit


In [3]:
from sklearn import datasets
#Cargamos el data set de boston
boston=datasets.load_boston()

In [7]:
from sklearn.linear_model import LinearRegression

In [8]:
#Creamos el modelo
model=LinearRegression()

In [9]:
#Ajustamos el modelo a los datos
model.fit(X=boston["data"],y=boston["target"])

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None,
         normalize=False)

In [15]:
y_objetivo=boston["target"]
#Prediccion del modelo con las variables independientes
y_pred=model.predict(boston["data"])

In [16]:
from sklearn import metrics

# Error Absoluto medio

El error absoluto medio(Mean Absolute Error o MAE)
$$\frac{1}{n}\sum_{i=1}^{n}|y_i -\hat{y}_i|$$

Es decir, la medida de las diferencias entre la variable objetivo y las predicciones sin signo.

MAE es una metrica robusta, en cuanto a que no varia mucho si hay valores extremos en los datos. El error se puede interpretar como unidades de la variable objetivo(por ejemplo, si la variable objetivo es en dolares MAE estara tambien en dolares).

In [18]:
metrics.mean_absolute_error?

In [19]:
metrics.mean_absolute_error(y_objetivo,y_pred)

3.2708628109003177

Nos arroja un error de mas menos 3.27.. mil dolares los precios 
de las casas en boston

# Error cuadratico Medio
El error cuadratico medio(Mean Squared Error o MSE)
$$\frac{1}{n}\sum_{i=1}^{n}(y_i -\hat{y}_i)^2$$

Dado que el MSE se define en unidades cuadraticas ,lo cual no es intituivo(¿dolares cuadrados?), generalmente se usa su raiz.

# Raíz del error cuadratico medio.
La raiz del error cuadrático medio( Root Mean Squared Error or RSME) se diferencia del MSE en que el resultado se puede medir en las mismas unidades de la variable objetivo.
$$\sqrt{\frac{1}{n}\sum_{i=1}^{n}(y_i -\hat{y}_i)^2}$$

Si embargo,  tiene un problema y es si es que da mas inportancia a los grandes errores. Por ejemplo en el Boston Housing Dataset, si tenemos dos observaciones y sus predicciones:

    observacion1: MEDV: 10 MEDV_pred: 15 RMSE:(10-15)^2 =25
    observacion2: MEDV: 1000 MEDV_pred: 1010 RMSE:(1000-1010)^2=100

El usar RMSE como medida de error significa que le dara mas peso al error de la observacion2 que al error de la observacion1, pese a que un error de 5000 dolares en una zona donde el valor medio es 10,000$ es un error mucho mas grave que un error de 10,000 dolares en una zona donde el valor medio de las casas es de un millon de dolares.

RMSE es mas sencible que MAE a variaciones en los errores de prediccion, aqui tenemos un ejemplo donde se ve esto.

![mae_vs_mse](https://cdn-images-1.medium.com/max/1371/1*YTxb8K2XZIisC944v6rERw.png)

Este es un ejemplo se ve como para los tres casos, el MAE no varía, mientras que en funcion de la distribucion de errores el RMSE puede ser igual al MAE(todos los errores iguales) o mucho mayor(muchos errores nulos con un erro muy grande).

Para aquellos casos donde lo que queramos es esto(evitar errores grandes) y cuando la distribucion de la variable objetivo este bien distribuida se puede usar RMSE en vez de MAE.

In [23]:
import numpy as np
np.sqrt(metrics.mean_squared_error(y_objetivo,y_pred))

4.679191295697281

# Coeficiente de determinacion
El coeficiente de determinacion(R2 o R-squared) mide la porcion de la varianza de la varible objetivo que se puede explicar por el modelo.

R2 tiene un nvalor maximo de 1(cuando el modelo explica toda la varianza), aunque puede tener valores negativos.

Hay varias formas de definir R2, pero una de las cosas mas sencillas es simplemente la correlacion(definida como la Correlacion de Pearson) entre la variable objetivo y las predicciones, elevada al cuadrado.

![pmcc](https://wikimedia.org/api/rest_v1/media/math/render/svg/01d103c10e6d4f477953a9b48c69d19a954d978a)

Un problema inportante que tiene R2 es que nos indica si el modelo explica la varianza debido a que esta sobreajusta(overfitted). Por eso una medida es el Coeficiente de Determinacion ajustado(Adjusted R.squared), que tiene en consideracion la conplejidad del mismo modelo.

$$1 - \frac{(1 - R^2)(n-1)}{(n-k-1)}$$

donde n es el numero de observaciones y k es el numero de coeficientes del modelo(sin contar el termino independiente)


In [25]:
modelo_r2=metrics.r2_score(y_objetivo,y_pred)
modelo_r2

0.7406426641094095

In [26]:
import numpy as np
np.corrcoef(y_objetivo,y_pred)**2

array([[1.        , 0.74064266],
       [0.74064266, 1.        ]])

Calcular el coeficiente de Determinacion ajustado

In [30]:
len(model.coef_)

13

In [29]:
modelo_r2_ajustado=1-(1-modelo_r2)*(len(boston["target"])-1)/\
                    (len(boston["target"])-boston["data"].shape[1]-1)
modelo_r2_ajustado

0.733789726372463